src/async_writer.cpp

```diff
@@ -32,18 +33,40 @@ void AsyncWriter::setConnectionAliveHandler(OnConnectionAliveCallback connection
     this->connectionAliveCallback = connectionAliveCallback;
 }
 
-void AsyncWriter::queueWriteTasks(std::vector<WriteTask>&& tasks)
+void AsyncWriter::setWriteTimedOutHandler(OnWriteTaskTimedOutCallback writeTaskTimeoutHandler)
+{
+    this->writeTaskTimeoutHandler = writeTaskTimeoutHandler;
+}
+
+void AsyncWriter::queueBatchWrite(BatchedWriteTasks&& tasks, OptionalWriteDeadline&& deadlineTime)
 {
     bool writing = !writeTasksQueue.empty();
-    writeTasksQueue.push(std::move(tasks));
+    writeTasksQueue.push(std::pair(std::move(tasks), std::move(deadlineTime)));
     if (!writing)
     {
         doWrite(writeTasksQueue.front());
     }
 }
 
-void AsyncWriter::doWrite(const std::vector<WriteTask>& tasks)
+void AsyncWriter::doWrite(const BatchedWriteTasksWithDeadline& tasksWithDeadline)
```

Comment:
At what point is doWrite triggered? When the previous task finishes or when a new packet is available? If it is triggered when the previous write task is finished, the task could potentially never finish, and therefore, the timeout would never be checked.

Reply:
If the batched write tasks queue is empty, meaning there are no pending write operations, then `doWrite` is triggered manually. Conversely, if previous write operation is still in progress when a new packet arrives, `doWrite` for that packet is triggered automatically after all previous operationd complete.

You’re correct - currently, the deadline is checked and the timeout error is triggered only on the previous write completion. So, if the previous write takes too long, the timeout error is reported late, and if it never completes, no timeout occurs at all. While this doesn’t address the second case (guess nothing really would help here as it is an extreme scenario where things go completely wrong), I’m considering adding a timer that starts in parallel with the async write operation in `doWrite`. This timer would trigger a timeout error if the write operation itself exceeds the deadline; and the timer be canceled if the write finishes in time.

As a result, `doWrite` should not initiate the next write operation if the deadline for that task has already passed (current behavior). Additionally, the timeout error could be reported from the timer callback if the write operation takes too long.

Reply:
The timer seems like a good solution to me!

However, if possible, I would use the following algorithm:
* Set `writeCompleted` to false
* Set `writePending` flag to true if there's any async write operation pending, otherwise set it to false. 
* Start the timer
* If write operation completes and `writePending` is true, set `writeCompleted` to true.
       - more write operations could be completed but that does not alter the behavior.
* When timer elapses
    - if `writePending == true` and `writeCompleted = false` -> we have a timeout error
    - go to the first  step
    
   This would allow to start the timer only once per `writeTimeout` and not once per `doWrite`, potentially never cancelling it.

Reply:
The final version works not exactly as you escribed, but I believe it  serves same purpose, if I understand your idea correctly. 

However, one potential overhead in this solution is that separate timer starts for each queued batch write with a deadline specified. The timer expires exactly at the specified deadline time, with the callback handled in the same context (thread) as the write completion callback. The the write completion callback `writeDone` being successful cancels and removes the timer along with the completed task from the queue. 

In my view, having a separate timer for each batch write (i.e., per each serialized openDAQ packet sent) seems like the only way to ensure the connection will be closed immediately on timeout for any streaming packet exceeds its deadline, independently of pending or previously completed writes or following queued write tasks.

Reply:
My only concern was the potential overhead of having a timer on each write. Hopefully, it is not too high.